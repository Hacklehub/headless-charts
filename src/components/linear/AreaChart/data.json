[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
