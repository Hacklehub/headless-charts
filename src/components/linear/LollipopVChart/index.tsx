import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { max, min } from 'd3-array';
import { scaleBand, scaleLinear } from 'd3';
import { select, selectAll } from 'd3-selection';
import {
  symbol,
  symbolCircle,
  symbolCross,
  symbolDiamond,
  symbolSquare,
  symbolStar,
  symbolTriangle,
  symbolWye,
} from 'd3';
import { useCallback, useEffect } from 'react';

import { defaultChartClassNames } from '../../../utils';
import { mergeTailwindClasses } from '../../../utils';

interface DataItem {
  [key: string]: any;
}

interface LollipopVChartProps {
  data: DataItem[];
  valueMin?: number;
  valueMax?: number;
  id: string;
  className?: string;
  classNamePoints?: string;
  classNameLines?: string;
  classNameSymbols?: string;
  margin?: {
    left: number;
    right: number;
    top: number;
    bottom: number;
  };
  padding?: {
    left: number;
    right: number;
    bottom: number;
    top: number;
  };
  shape:
    | 'circle'
    | 'diamond'
    | 'triangle'
    | 'square'
    | 'cross'
    | 'star'
    | 'wye';
  x?: { axis: 'bottom' | 'top'; axisTicks: number; key: string };
  y?: {
    axis?: 'left' | 'right';
    axisTicks?: number;
    key: string;
    start?: number;
  };
}

const LollipopVChart = ({
  data = [],
  //   valueMin,
  valueMax,
  id,
  className,
  classNamePoints,
  classNameLines,
  classNameSymbols,
  margin = {
    left: 40,
    right: 40,
    top: 40,
    bottom: 40,
  },
  padding = {
    left: 0,
    right: 0,
    bottom: 0,
    top: 0,
  },
  shape = 'circle',
  x = { axis: 'bottom', axisTicks: 0, key: 'x' },
  y = { axis: 'left', key: 'y' },
}: LollipopVChartProps) => {
  const refreshChart = useCallback(() => {
    const svg = select(`#${id}`);
    // Clear svg
    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0];
    const height = +svg.style('height').split('px')[0];

    data.sort((a, b) => b[y.key] - a[y.key]);

    const shapeMapping = {
      circle: symbolCircle,
      diamond: symbolDiamond,
      triangle: symbolTriangle,
      square: symbolSquare,
      cross: symbolCross,
      star: symbolStar,
      wye: symbolWye,
    };

    const minValue = Number.isFinite(y.start)
      ? y.start
      : min(data, (d) => d[y.key]);
    const maxValue = Number.isFinite(valueMax)
      ? valueMax
      : max(data, (d) => d[y.key]);

    const xFn = scaleBand()
      .domain(data.map((d) => d[x.key]))
      .range([
        margin.left + padding.left,
        width - margin.right - padding.right,
      ]);

    const yFn = scaleLinear()
      .domain([minValue, maxValue])
      .range([
        height - margin.bottom - padding.bottom,
        margin.top + padding.top,
      ]);

    const g = svg.append('g');

    const xAxis = x.axis === 'top' ? axisTop(xFn) : axisBottom(xFn);

    const xAxisG = g.append('g').attr('class', 'axis--x axis');

    xAxisG
      .attr(
        'transform',
        `translate(0, ${
          x.axis === 'top' ? margin.top : height - margin.bottom
        })`
      )
      .transition()
      .duration(1000)
      .call(xAxis);

    const yAxis =
      y.axis === 'right'
        ? axisRight(yFn).ticks(y.axisTicks || 2)
        : axisLeft(yFn).ticks(y.axisTicks || 2);

    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${y.axis === 'right' ? margin.left + width : margin.left},0)`
      );

    padding.left &&
      xAxisG
        .append('line')
        .attr('x1', 0)
        .attr('x2', padding.left)
        .attr('y1', 0)
        .attr('y2', 0)
        .attr('stroke', 'currentColor');

    padding.bottom &&
      yAxisG
        .append('line')
        .attr('x1', 0)
        .attr('x2', 0)
        .attr('y1', margin.top + height - padding.bottom)
        .attr('y2', margin.top + height)
        .attr('stroke', 'currentColor');

    yAxisG.transition().duration(1000).call(yAxis);

    const dataGroup = g.append('g');
    const drawLinesAndCircles = () => {
      const pointGroup = dataGroup
        .selectAll('.line')
        .data(data)
        .enter()
        .append('g');

      /* eslint-disable */
      pointGroup
        .append('line')
        .attr(
          'class',
          `line stroke-current ${classNamePoints || ''} ${classNameLines || ''}`
        )
        // @ts-ignore
        .attr('x1', (d) => xFn(d[x.key]) + xFn.bandwidth() / 2)
        .attr('y1', () => yFn(minValue))
        // @ts-ignore
        .attr('x2', (d) => xFn(d[x.key]) + xFn.bandwidth() / 2)
        .attr('y2', () => yFn(minValue))
        .transition()
        .duration(1000)
        .attr('y1', (d) => yFn(d[y.key]));

      pointGroup
        .append('path')
        .attr(
          'class',
          `symbols fill-current ${classNamePoints || ''} ${
            classNameSymbols || ''
          }`
        )
        .attr('d', () => symbol(shapeMapping[shape], 100)())
        .attr(
          'transform',
          (d) =>
            //@ts-ignore
            `translate(${xFn(d[x.key]) + xFn.bandwidth() / 2},${yFn(minValue)})`
        )
        .transition()
        .duration(1000)
        .attr(
          'transform',
          (d) =>
            // @ts-ignore
            `translate(${xFn(d[x.key]) + xFn.bandwidth() / 2},${yFn(
              d[y.key]
            )} )`
        );
    };
    /* eslint-enable */
    drawLinesAndCircles();
  }, [
    classNameLines,
    classNamePoints,
    classNameSymbols,
    data,
    id,
    shape,
    valueMax,
    x,
    y,
    margin,
    padding,
  ]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll('.tooltip').remove();
    };
  }, [data, refreshChart]);

  return (
    <svg
      id={id}
      className={mergeTailwindClasses(defaultChartClassNames, className)}
    />
  );
};

export default LollipopVChart;
