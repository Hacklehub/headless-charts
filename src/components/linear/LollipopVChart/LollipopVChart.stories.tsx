import LollipopVChart from '.';
import { Meta } from '@storybook/react';
import data from './sample.json';

export default {
  title: 'Linear/LollipopVChart',
  component: LollipopVChart,
  tags: ['autodocs'],
} as Meta;

export const Default = {
  args: {
    data,
    id: 'lollipop-v-chart-default',
    x: {
      axis: 'bottom',
      axisTicks: 2,
      key: 'name',
    },
    y: {
      axis: 'left',
      axisTicks: 4,
      key: 'reading',
      start: 0,
    },
  },
};

export const WithCustomShape = {
  args: {
    ...Default.args,
    shape: 'star',
  },
};

export const WithCustomStyles = {
  args: {
    ...WithCustomShape.args,
    classNameLines: 'fill-red-500 stroke-red-500',
    classNameSymbols: 'fill-blue-500 stroke-blue-500',
  },
};
