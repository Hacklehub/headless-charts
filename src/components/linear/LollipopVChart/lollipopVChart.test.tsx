import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import LollipopVChart from '.';
import data from './sample.json';

describe('LollipopVChart', () => {
  it('renders lollipopVChart component for snapshot test', () => {
    render(<LollipopVChart id='lollipopHChart' data={data} shape='star' />);
    const lollipopVChart = screen.getByTestId('lollipopVChart');
    expect(lollipopVChart).toMatchSnapshot();
  });
  it('renders lollipopHChart component', () => {
    render(<LollipopVChart id='lollipopHChart' data={data} shape='star' />);
    const lollipopVChart = screen.getByTestId('lollipopVChart');
    expect(lollipopVChart).toBeInTheDocument();
  });
});
