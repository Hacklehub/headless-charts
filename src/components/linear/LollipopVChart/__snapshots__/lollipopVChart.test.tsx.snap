// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LollipopVChart > renders lollipopVChart component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="lollipopVChart"
  id="lollipopHChart"
>
  <g>
    <g
      class="axis--x axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        />
      </g>
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        stroke="currentColor"
      />
    </g>
    <g>
      <g>
        <line
          class="line stroke-current  "
          x1="0"
          x2="0"
        />
        <path
          class="symbols fill-current  "
          d="M0,-9.438L2.119,-2.917L8.976,-2.917L3.429,1.114L5.548,7.636L0,3.605L-5.548,7.636L-3.429,1.114L-8.976,-2.917L-2.119,-2.917Z"
          transform="translate(0,undefined)"
        />
      </g>
      <g>
        <line
          class="line stroke-current  "
          x1="0"
          x2="0"
        />
        <path
          class="symbols fill-current  "
          d="M0,-9.438L2.119,-2.917L8.976,-2.917L3.429,1.114L5.548,7.636L0,3.605L-5.548,7.636L-3.429,1.114L-8.976,-2.917L-2.119,-2.917Z"
          transform="translate(0,undefined)"
        />
      </g>
    </g>
  </g>
</svg>
`;
